dbms/src/Interpreters/Join.h

```diff
@@ -151,22 +151,68 @@ class Join
         RowRefList(const Block * block_, size_t row_num_) : RowRef(block_, row_num_) {}
     };
 
-    /// Map for a time series
-    using ASOFTimeType = UInt32;
-    using AsofGetterType = ColumnsHashing::HashMethodOneNumber<ASOFTimeType, ASOFTimeType, ASOFTimeType, false>;
-    struct TSRowRef
+    struct AsofRowRefs
     {
-        // TODO use the arena allocator to get memory for this
-        // This would require ditching std::map because std::allocator is incompatible with the arena allocator
-        std::map<ASOFTimeType, RowRef> ts;
-
-        TSRowRef() {}
-        void insert(ASOFTimeType t, const Block * block, size_t row_num);
-        std::optional<std::pair<ASOFTimeType, RowRef>> findAsof(ASOFTimeType t) const;
-        std::string dumpStructure() const;
-        size_t size() const;
+        /// Different types of asof join keys
+        #define APPLY_FOR_ASOF_JOIN_VARIANTS(M) \
+            M(key32, UInt32)                    \
+            M(key64, UInt64)                    \
+            M(keyf32, Float32)                  \
+            M(keyf64, Float64)
+
+        enum class AsofType
+        {
+            EMPTY,
+        #define M(NAME, TYPE) NAME,
+            APPLY_FOR_ASOF_JOIN_VARIANTS(M)
+        #undef M
+        };
+
+        static std::optional<std::pair<AsofType, size_t>> getTypeSize(const IColumn * asof_column);
+
+        template<typename T>
+        struct Entry
+        {
+            T asof_value;
+            RowRef row_ref;
+
+            Entry(T v) : asof_value(v) {}
+            Entry(T v, RowRef rr) : asof_value(v), row_ref(rr) {}
+
+            bool operator< (const Entry& o) const
+            {
+                return asof_value < o.asof_value;
+            }
+        };
+
+        struct Lookups
+        {
+            #define M(NAME, TYPE) \
+            std::unique_ptr<PODArray<Entry<TYPE>>> NAME;
```

Comment:
Could we make more general thing like SortedLookupPODArray and place it in Common instead of this?
We could base on PaddedPODArry and add sort(), upper_bound(), lower_bound() methods.